File: src/capture.py
import threading
import time
from pathlib import Path

import cv2


class ThreadedCapture:
    """
    Capture frames in a background thread and always
    expose the most recent frame.

    Advantages on a weak board:

    - Camera capture never blocks inference.
    - Stale frames are dropped instead of queued.
    - A single reusable frame buffer avoids copies.
    """

    def __init__(
        self,
        source,
        width=None,
        height=None,
        fps=None,
        use_mjpeg=True,
        buffersize=1,
    ):
        """
        Args:
            source:
                Camera index (int) or path to a video file.
            width, height, fps:
                Optional capture properties (camera only).
            use_mjpeg:
                Request MJPG format from the camera.
                Much cheaper to decode on weak CPUs.
            buffersize:
                Internal frame queue size (keep at 1 for low latency).
        """

        self._source = source
        self._width = width
        self._height = height
        self._fps = fps
        self._use_mjpeg = use_mjpeg
        self._buffersize = buffersize

        self._capture = None
        self._frame = None
        self._finished = False
        self._lock = threading.Lock()
        self._running = False
        self._thread = None


    def start(self):
        """Open the source and start the capture thread."""

        if isinstance(self._source, (str, Path)):

            self._capture = cv2.VideoCapture(
                str(self._source)
            )

        else:

            self._capture = cv2.VideoCapture(
                int(self._source)
            )


        if not self._capture.isOpened():
            raise RuntimeError(
                f"Could not open video source: {self._source}"
            )


        if isinstance(self._source, int):

            if self._use_mjpeg:

                self._capture.set(
                    cv2.CAP_PROP_FOURCC,
                    cv2.VideoWriter_fourcc(*"MJPG")
                )

            if self._width:
                self._capture.set(
                    cv2.CAP_PROP_FRAME_WIDTH,
                    int(self._width)
                )

            if self._height:
                self._capture.set(
                    cv2.CAP_PROP_FRAME_HEIGHT,
                    int(self._height)
                )

            if self._fps:
                self._capture.set(
                    cv2.CAP_PROP_FPS,
                    float(self._fps)
                )

            self._capture.set(
                cv2.CAP_PROP_BUFFERSIZE,
                int(self._buffersize)
            )

            self._file_delay = None

        else:

            # Throttle video playback to its native speed so
            # the whole file does not finish instantly.

            playback_fps = (
                self._capture.get(cv2.CAP_PROP_FPS)
            )

            if playback_fps and playback_fps > 0:
                self._file_delay = 1.0 / playback_fps
            else:
                self._file_delay = None


        self._running = True

        self._thread = threading.Thread(
            target=self._loop,
            daemon=True
        )

        self._thread.start()


    def _loop(self):
        """Background capture loop."""

        while self._running:

            success, frame = self._capture.read()

            if not success:

                # A video has ended; stop the thread.
                # A camera had a hiccup; keep trying.

                if isinstance(self._source, (str, Path)):
                    self._finished = True
                    break

                continue


            with self._lock:
                self._frame = frame


            if self._file_delay:
                time.sleep(self._file_delay)


    def read(self):
        """
        Return the most recent frame.

        Returns:
            Latest frame, or None if no frame is ready yet
            (or the video file has ended).
        """

        if self._finished:
            return None

        with self._lock:
            frame = self._frame

        return frame


    def is_file(self):
        return isinstance(
            self._source,
            (str, Path)
        )

    def finished(self):
        """True when a video file has reached the end."""

        return self._finished


    def stop(self):
        """Stop the capture thread and release the source."""

        self._running = False

        if self._thread:
            self._thread.join(timeout=1)


        if self._capture:
            self._capture.release()

File: cpp/src/capture.cpp
#include "capture.hpp"

#include <chrono>
#include <stdexcept>

namespace tp {

ThreadedCapture::ThreadedCapture(int source,
                                 int width, int height, double fps,
                                 bool use_mjpeg, int buffersize)
    : source_index_(source),
      is_file_source_(false),
      width_(width),
      height_(height),
      fps_(fps),
      use_mjpeg_(use_mjpeg),
      buffersize_(buffersize) {}

ThreadedCapture::ThreadedCapture(const std::string& file_path, bool use_mjpeg)
    : file_path_(file_path),
      is_file_source_(true),
      use_mjpeg_(use_mjpeg),
      buffersize_(1) {}

ThreadedCapture::~ThreadedCapture() {
    stop();
}

void ThreadedCapture::start() {
    const bool opened = is_file_source_ ? capture_.open(file_path_)
                                        : capture_.open(source_index_);
    if (!opened) {
        throw std::runtime_error(
            "Could not open video source: " +
            (is_file_source_ ? file_path_ : std::to_string(source_index_)));
    }

    if (!is_file_source_) {
        if (use_mjpeg_) {
            capture_.set(cv::CAP_PROP_FOURCC,
                         cv::VideoWriter::fourcc('M', 'J', 'P', 'G'));
        }
        if (width_) {
            capture_.set(cv::CAP_PROP_FRAME_WIDTH, width_);
        }
        if (height_) {
            capture_.set(cv::CAP_PROP_FRAME_HEIGHT, height_);
        }
        if (fps_ > 0.0) {
            capture_.set(cv::CAP_PROP_FPS, fps_);
        }
        capture_.set(cv::CAP_PROP_BUFFERSIZE, buffersize_);
        has_file_delay_ = false;
    } else {
        // Throttle video playback to its native speed so the whole
        // file does not finish instantly.
        const double playback_fps = capture_.get(cv::CAP_PROP_FPS);
        if (playback_fps > 0.0) {
            file_delay_ = 1.0 / playback_fps;
            has_file_delay_ = true;
        } else {
            has_file_delay_ = false;
        }
    }

    running_ = true;
    thread_ = std::thread(&ThreadedCapture::loop, this);
}

void ThreadedCapture::loop() {
    while (running_) {
        cv::Mat frame;
        const bool success = capture_.read(frame);
        if (!success) {
            // A video has ended; stop the thread.
            // A camera had a hiccup; keep trying.
            if (is_file_source_) {
                {
                    std::lock_guard<std::mutex> lock(mutex_);
                    finished_ = true;
                }
                break;
            }
            continue;
        }

        {
            std::lock_guard<std::mutex> lock(mutex_);
            frame_ = frame;
        }

        if (has_file_delay_) {
            std::this_thread::sleep_for(
                std::chrono::duration<double>(file_delay_));
        }
    }
}

cv::Mat ThreadedCapture::read() {
    std::lock_guard<std::mutex> lock(mutex_);
    if (finished_) {
        return cv::Mat();
    }
    return frame_.clone();
}

bool ThreadedCapture::is_file() const {
    return is_file_source_;
}

bool ThreadedCapture::finished() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return finished_;
}

void ThreadedCapture::stop() {
    running_ = false;
    if (thread_.joinable()) {
        thread_.join();
    }
    if (capture_.isOpened()) {
        capture_.release();
    }
}

}  // namespace tp